src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +113,246 @@ object RuleUtils {
       None // logicalPlan is non-linear or it has no LogicalRelation.
     }
   }
+
+  /**
+   * Get replacement plan for current plan. The replacement plan reads data from indexes.
+   * If HybridScan is enabled, additional logical plans for the appended data would be
+   * generated and merged with index data plan. Refer [[getComplementIndexPlan]].
+   *
+   * Pre-requisites
+   * - We know for sure the index which can be used to replace the plan.
+   *
+   * NOTE: This method currently only supports replacement of Scan Nodes i.e. Logical relations
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Replacement plan.
+   */
+  def getReplacementPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    if (HyperspaceConf.hybridScanEnabled(spark)) {
+      getHybridScanIndexPlan(spark, index, plan, useBucketSpec)
+    } else {
+      getIndexPlan(spark, index, plan, useBucketSpec)
+    }
+  }
+
+  /**
+   * Get alternative logical plan of the current plan using the given index.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Alternative logical plan.
+   */
+  private def getIndexPlan(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    // Here we can't replace the plan completely with the index. This will create problems.
+    // For e.g. if Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
+    // if we replace this plan with index Scan (A,B,C), we lose the Filter(C=10) and it will
+    // lead to wrong results. So we only replace the base relation.
```

Comment:
```suggestion
    // Note that we replace *only* the base relation and not any sub plans (e.g., filters).
    // For instance, given the following input plan:
    //        Project(A,B) -> Filter(C = 10) -> Scan (A,B,C,D,E)
    // in the presence of a suitable index, this method will emit:
    //        Project(A,B) -> Filter(C = 10) -> Index Scan (A,B,C)
    // In other words, the Filter is untouched to preserve correctness. 
```

Reply:
One question about the line length limit. It seems our line length limit is `maxColumn = 98` (in scalafmt.conf) though the below coding style guides recommend 80 and 100. Is there any reason for using 98?

https://docs.scala-lang.org/style/indentation.html#line-wrapping => 80
https://github.com/databricks/scala-style-guide#linelength => 100

Reply:
Good question. :) It was inspired from [Spark's scalafmt.conf](https://github.com/apache/spark/blob/master/dev/.scalafmt.conf)